src/FsAutoComplete/LspServers/FSharpLspClient.fs

```diff
@@ -143,74 +144,164 @@ type ServerProgressReport(lspClient: FSharpLspClient, ?token: ProgressToken) =
+            // We don't seem to get a corresponding ActivityStopped on cancelled typechecks.
+            if isStopped a then
+              do! p.End()
+              inflightEvents.TryRemove(a.Id) |> ignore
+            else
+              // FSC doesn't start their spans with tags so we have to see if it's been added later https://github.com/dotnet/fsharp/issues/14776
+              let fileName = getFileName a
+              let userOpName = getUserOpName a
+              do! p.Report(message = $"{fileName} - {userOpName}")
+
+          match! inbox.TryReceive(250) with
+          | None ->
+            // if we don't get an event in 250 ms just loop again so we can analyze inflightEvents
+            ()
+          | Some(action, activity: Activity, reply: AsyncReplyChannel<unit>) ->
+
+            match action with
+            | "start" ->
+              let fileName = getFileName activity
+              let userOpName = getUserOpName activity
+
+              logger.debug (
+                Log.setMessageI
+                  $"Started : {activity.DisplayName:DisplayName} - {userOpName:UserOpName} - {fileName:fileName}"
+              )
+
+              if
+                activity.DisplayName |> isOneOf interestingActivities
+                && not (isStopped activity)
+              then
+                let progressReport = new ServerProgressReport(lspClient)
+
+                if inflightEvents.TryAdd(activity.Id, (activity, progressReport)) then
+
+                  do! progressReport.Begin($"{activity.DisplayName}")
+
+            | "stop" ->
+              let fileName = getFileName activity
+              let userOpName = getUserOpName activity
+              let duration = activity.Duration.ToString()
+
+              logger.debug (
+                Log.setMessageI
+                  $"Finished : {activity.DisplayName:DisplayName} - {userOpName:UserOpName} - {fileName:fileName} - took {duration:duration}"
+              )
+
+              if activity.DisplayName |> isOneOf interestingActivities then
+                match inflightEvents.TryRemove(activity.Id) with
+                | true, (old, progressReport) -> do! progressReport.End()
+                | _ -> ()
+
+            | _ -> ()
+
+            reply.Reply()
+      })
+
+
+  let shouldListenTo (act: ActivitySource) = act.Name = Tracing.fscServiceName
+
+  let activityStarted (act: Activity) =
+    mbp.PostAndReply(fun reply -> "start", act, reply)
+
+  let activityStopped (act: Activity) =
+    mbp.PostAndReply(fun reply -> "stop", act, reply)
+
+  let listener =
+    new ActivityListener(
+      ShouldListenTo = shouldListenTo,
+      Sample = (fun _ -> ActivitySamplingResult.AllDataAndRecorded),
+      ActivityStarted = activityStarted,
+      ActivityStopped = activityStopped
+    )
+
+  do ActivitySource.AddActivityListener listener
```

Comment:
We discussed this being able to be enabled with a setting, still want to do that?

Reply:
yeah - if you could make it toggleable that would be the last thing. otherwise I'm super happy, this looks amazing.

Reply:
done